### PR TITLE
executor: fix a panic when batch point get is used for partition table

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve?

5.0.0-rc panic stack:

```
{"level":"INFO","time":"2021/03/25 10:54:26.757 +08:00","caller":"conn.go:809","message":"command dispatched failed","conn":49,"connInfo":"id:49, addr:11.186.250.167:57210 status:10, collation:utf8mb4_general_ci, user:media_inter_r","command":"Execute","status":"inTxn:0, autocommit:1","sql":"SELECT `imp_date`, `media_id`, `app`, `article_likepv`, `article_collectpv`, `article_relaypv`, `article_commentpv`, `video_likevv`, `video_collectvv`, `video_relayvv`, `video_commentvv` FROM `media_interact_daily_6`.`t_media_interact_daily_4` WHERE (media_id=? AND app=? AND ((imp_date>=? AND imp_date<=?) OR (imp_date>=? AND imp_date<=?) OR (imp_date>=? AND imp_date<=?) OR (imp_date>=? AND imp_date<=?))) [arguments: (19790046, all, 2021-03-24, 2021-03-24, 2021-03-23, 2021-03-23, 2021-03-17, 2021-03-17, 2021-02-22, 2021-02-22)]","txn_mode":"OPTIMISTIC","err":"runtime error: integer divide by zero
github.com/pingcap/tidb/executor.(*recordSet).Next.func1
	/home/jenkins/agent/workspace/build-linux-amd64-enterprise/go/src/github.com/pingcap/tidb/executor/adapter.go:125
runtime.gopanic
	/usr/local/go/src/runtime/panic.go:679
runtime.panicdivide
	/usr/local/go/src/runtime/panic.go:178
github.com/pingcap/tidb/executor.getPhysID
	/home/jenkins/agent/workspace/build-linux-amd64-enterprise/go/src/github.com/pingcap/tidb/executor/batch_point_get.go:427
github.com/pingcap/tidb/executor.(*BatchPointGetExec).initialize
	/home/jenkins/agent/workspace/build-linux-amd64-enterprise/go/src/github.com/pingcap/tidb/executor/batch_point_get.go:195
github.com/pingcap/tidb/executor.(*BatchPointGetExec).Next
	/home/jenkins/agent/workspace/build-linux-amd64-enterprise/go/src/github.com/pingcap/tidb/executor/batch_point_get.go:143
github.com/pingcap/tidb/executor.Next
	/home/jenkins/agent/workspace/build-linux-amd64-enterprise/go/src/github.com/pingcap/tidb/executor/executor.go:278
github.com/pingcap/tidb/executor.(*recordSet).Next
	/home/jenkins/agent/workspace/build-linux-amd64-enterprise/go/src/github.com/pingcap/tidb/executor/adapter.go:129
github.com/pingcap/tidb/server.(*tidbResultSet).Next
	/home/jenkins/agent/workspace/build-linux-amd64-enterprise/go/src/github.com/pingcap/tidb/server/driver_tidb.go:298
github.com/pingcap/tidb/server.(*clientConn).writeChunks
	/home/jenkins/agent/workspace/build-linux-amd64-enterprise/go/src/github.com/pingcap/tidb/server/conn.go:1732
github.com/pingcap/tidb/server.(*clientConn).writeResultset
	/home/jenkins/agent/workspace/build-linux-amd64-enterprise/go/src/github.com/pingcap/tidb/server/conn.go:1692
github.com/pingcap/tidb/server.(*clientConn).handleStmtExecute
	/home/jenkins/agent/workspace/build-linux-amd64-enterprise/go/src/github.com/pingcap/tidb/server/conn_stmt.go:213
github.com/pingcap/tidb/server.(*clientConn).dispatch
	/home/jenkins/agent/workspace/build-linux-amd64-enterprise/go/src/github.com/pingcap/tidb/server/conn.go:1050
github.com/pingcap/tidb/server.(*clientConn).Run
	/home/jenkins/agent/workspace/build-linux-amd64-enterprise/go/src/github.com/pingcap/tidb/server/conn.go:792
github.com/pingcap/tidb/server.(*Server).onConn
	/home/jenkins/agent/workspace/build-linux-amd64-enterprise/go/src/github.com/pingcap/tidb/server/server.go:461
runtime.goexit
	/usr/local/go/src/runtime/asm_amd64.s:1357
```

Problem Summary:

The '[divide by zero](https://github.com/pingcap/tidb/blob/fd4437dde445012f5b75e5d4f336433cb076c1ee/executor/batch_point_get.go#L427)' error should not happen, and in fact, partition table should not use BatchPointGet executor in the physical plan phase.

The root cause should be our improper handle of the partition meta info.
For historical reason, the meta information of a partition table maybe not nil, but its `Enable` field is false. 
That means it's not really a partition table. 
We should use `GetPartitionInfo()` check is a table a real partition table or not.

### What is changed and how it works?

What's Changed:

Change direct access of struct field `tableInfo.Partition` to `tableInfo.GetPartition()`

How it Works:

[GetPartitionInfo()](https://github.com/pingcap/parser/blob/ae4372a9539acc60fc66e074fe9410580496be92/model/model.go#L450) will check the `Enable` field, we should always use it instead of using `tableInfo.Partition` .

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test

### Release note 

- Fix a panic on batch point get for non-partitioned table with partition meta information. When a cluster is upgrade from a old version, its partition meta information maybe not null, but the `Enable` field is false, it should be treat as a non-partitioned table.

### CODE REVIEW

File: executor/partition_table_test.go [hunk 1]
```diff
@@ -176,6 +178,48 @@ PRIMARY KEY (pk1,pk2)) partition by hash(pk2) partitions 4;`)
...
+	c.Assert(err, IsNil)
+
+	tbInfo := tbl.Meta()
+	// Mock for a case that the tableInfo.Partition is not nil, but tableInfo.Partition.Enable is false.
+	// That may happen when upgrading from a old version TiDB.
+	tbInfo.Partition.Enable = false
+	tbInfo.Partition.Num = 0
+
+	tk.MustExec("set @@tidb_partition_prune_mode = 'static'")
+	tk.MustQuery("explain select * from t_info_null where (date = '2020-10-02' or date = '2020-10-06') and app = 'xxx' and media = '19003006'").Check(testkit.Rows("Batch_Point_Get_5 2.00 root table:t_info_null, index:idx_media_id(media, date, app) keep order:false, desc:false"))
```
Comment (djshow832) [inline executor/partition_table_test.go:218]: And `explain select * from t_info_null`. It shouldn't include `PartitionUnion`.

**Review by djshow832 — COMMENTED** [review]
Does any modification need to be cherry-picked to 4.0?

**Comment by tiancaiamao** [comment]
> Does any modification need to be cherry-picked to 4.0?

Oh, yes, 4.0 also use this function:

```
func getPhysID(tblInfo *model.TableInfo, val int64) int64 {
	pi := tblInfo.Partition
	if pi == nil {
		return tblInfo.ID
	}
	partIdx := math.Abs(val % int64(pi.Num))
	return pi.Definitions[partIdx].ID
}
```

File: executor/partition_table_test.go (hunk 1 above)
Comment (tiancaiamao) [reply executor/partition_table_test.go:218]: done

**Review by qw4990 — APPROVED** [review]

**Pull request merged**